custom_components/amplipi/media_players/AmpliPiStream.py

```diff
@@ -0,0 +1,254 @@
+            return STATE_IDLE
+        elif self._source.info.state in (
+                'paused'
+        ):
+            return STATE_PAUSED
+        elif self._source.info.state in (
+                'playing'
+        ):
+            return STATE_PLAYING
+        elif self._source.info.state in (
+                'stopped'
+        ):
+            return STATE_IDLE
+        return STATE_IDLE
+
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        if self._source is not None:
+            if len(self._zones) > 0:
+                vols = [zone.vol_f for zone in self._zones]
+                return sum(vols) / len(vols)
+        return None
+
+    @property
+    def is_volume_muted(self) -> bool:
+        """Boolean if all connected zones are currently muted. If no zones connected, return muted."""
+        if self._source is not None:
+            for zone in self._zones:
+                if not zone.mute:
+                    return False
+        return True
+ 
+    async def async_select_source(self, source: Optional[str] = None):
+        # This is a home assistant MediaPlayer built-in function, so the source being passed in isn't the same as an amplipi source
+        # the argument "source" can either be the name or entity_id of a zone, group, or amplipi source
+        # As such, this info must be sorted and then sent down the proper logical path
+        if source:
+            if source == "None" and self._source is not None:
+                await self._data_client.set_source(
+                    self._source.id,
+                    SourceUpdate(
+                        input='None'
+                    )
+                )
+            elif source == "Any" and self._source is None:
+                await self.async_connect_stream_to_source(self._stream)
+            else:
+                amplipi_entity = self.get_entry_by_value(source)
+                if isinstance(amplipi_entity, Source):
+                    await self.async_connect_stream_to_source(self._stream, amplipi_entity)
+                else:
+                    entry = self.get_entry_by_value(source)
+                    if entry:
+                        entry_id = self.extract_amplipi_id_from_unique_id(entry.unique_id)
+                        if isinstance(amplipi_entity, Zone):
+                            await self.async_connect_zones_to_stream(self._stream, [entry_id], None)
+                        elif isinstance(amplipi_entity, Group):
+                            await self.async_connect_zones_to_stream(self._stream, None, [entry_id])
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return self._attr_source_list
+
+    @property
+    def extra_state_attributes(self):
+        return {"stream_type" : self._stream.type}
+
+    async def _update_available(self):
+        if self._stream is None:
+            return False
+        return True
+    
```

Comment:
Looks like there's some trailing space here on the newline. We really need to add linting to the GH actions. Does this repo have an editor config setup?

Reply:
There one last one to fix on utils.py